src/bucket/BucketList.cpp

```diff
@@ -551,22 +571,91 @@ BucketList::addBatch(Application& app, uint32_t currLedger,
+BucketList::restartMerges(Application& app, uint32_t maxProtocolVersion,
+                          uint32_t ledgerSeq, bool useFutureBucketsFromHAS)
 {
+    std::shared_ptr<Bucket> snap;
     for (uint32_t i = 0; i < static_cast<uint32>(mLevels.size()); i++)
     {
         auto& level = mLevels[i];
         auto& next = level.getNext();
-        if (next.hasHashes() && !next.isLive())
+
+        // Determine whether to use future bucket from HAS, and restart merge
+        // with curr and snap for this level
+        auto removeShadows = spillLevelWithoutShadows(maxProtocolVersion, i);
+
+        if (!removeShadows || useFutureBucketsFromHAS)
         {
-            next.makeLive(app, maxProtocolVersion, i);
-            if (next.isMerging())
+            if (next.hasHashes() && !next.isLive())
             {
-                CLOG(INFO, "Bucket")
-                    << "Restarted merge on BucketList level " << i;
+                next.makeLive(app, maxProtocolVersion, i);
+                if (next.isMerging())
+                {
+                    CLOG(INFO, "Bucket")
+                        << "Restarted merge on BucketList level " << i;
+                }
             }
         }
+        else
+        {
+            assert(next.isClear());
+            if (i > 0)
+            {
+                // Recover merge by iterating through bucketlist levels and
+                // using snaps and currs. The only time we don't use level's
+                // curr is when a level is about to snap; in that case, when the
+                // merge is needed, level's curr will snap, and merge will be
+                // promoted into curr. Therefore, when starting merge, we use an
+                // empty curr.
+                // Additionally, it is safe to recover a merge at any point
+                // before the merge is needed (meaning it should be promoted
+                // into level's curr after ledger close). This is due to the
+                // fact that future bucket inputs _do not change_ until level
+                // spill, and after such spills, new merges are started with new
+                // inputs.
+                auto lastMergeLedger = ledgerSeq /
+                                       BucketList::levelHalf(i - 1) *
+                                       BucketList::levelHalf(i - 1);
+                uint32_t nextSpillLedger =
+                    lastMergeLedger + BucketList::levelHalf(i - 1);
+                auto shouldClearCurr =
+                    BucketList::levelShouldSpill(nextSpillLedger, i);
+                auto newCurr = shouldClearCurr ? std::make_shared<Bucket>()
+                                               : level.getCurr();
+                bool countMergeEvents =
+                    !app.getConfig()
+                         .ARTIFICIALLY_REDUCE_MERGE_COUNTS_FOR_TESTING;
+                auto fb = FutureBucket(app, newCurr, snap,
```

Comment:
I still feel like carrying `snap` forward from one iteration of this loop to the next is a bit too non-obvious -- at least it's surprised me both times I've reviewed this loop! -- and I think it's only used in this expression here, right? Would it be possible to put the expression `mLevels[i-1].getSnap()` here, and remove the loop-carried variable `snap`?

Reply:
fixed